util/strong_random.py

```diff
@@ -0,0 +1,137 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+"""Random number generators.
+
+A class for generating random numbers and permutations, by fetching entropy
+from a file.
+"""
+
+import random as rnd
+from math import ceil as _ceil
+from math import log as _log
+from warnings import warn as _warn
+
+import numpy as np
+
+
+class strong_random():
+
+    def __init__(self):
+        """Initialize an instance.
+
+        Entropy buffer is empty by default.
+        """
+
+        self.next = None
+        self.empty = True
+        self.buffer = np.array([])
+
+    def load(self, input_file):
+        """Load entropy buffer from an external file.
+
+        Currently only supports numpy array of 8-bit values.
+        """
+
+        self.buffer = np.load(input_file)
+        self.empty = False
+        self.next = 0
+
+    def size(self):
+        """ Size of the buffer.
+        """
+        return np.size(self.buffer)
+
+    def printstatus(self):
+        """Print the status of the buffer and the number of remaining bytes.
+
+        May be useful for testing and debugging.
+        """
+
+        print(self.buffer)
+        if self.empty:
+            print('Buffer is empty.')
+        else:
+            print("Buffer remaining:", len(self.buffer), "bytes.")
+
+    def isempty(self):
+        """Checks if the buffer is empty.
+        """
+        return self.empty
+
+    def fetchbyte(self):
+        """Fetch a random byte from the buffer.
+
+        Used by getrandbits().
+        Fetches a random byte from a buffer and updates the self.empty status.
+        If the buffer is already empty, issues a warning and defaults to
+        pseudo-random numbers from random.py.
+        """
+
+        if self.empty:
+            _warn('Buffer empty, defaulting to random.py', UserWarning)
+            random_byte = rnd.getrandbits(8)
+            return (random_byte)
+        else:
+            random_byte = self.buffer[0]
+            self.buffer = np.delete(self.buffer, 0)
+            self.empty = (self.buffer.size == 0)
+            return (random_byte)
+
+    def getrandbits(self, n_bits):
+        """Fetch n_bits random bits from a buffer.
+
+        Always fetches a whole number of bytes. Unused bits are discarded.
+        """
+
+        bitsleft = n_bits
+        R = 0
+        while (bitsleft > 0):
+            random_byte = self.fetchbyte()
+            if (bitsleft > 8):
+                R = (R << 8) | random_byte
+                bitsleft = bitsleft - 8
+            else:
+                random_byte >>= 8 - bitsleft
+                R = (R << bitsleft) | random_byte
+                bitsleft = 0
+        return (R)
+
+    def randbellow(self, n):
```

Comment:
typo:
```suggestion
    def randbelow(self, n):
```

Reply:
Thanks for catching this. It's corrected now.